crates/validator/src/server/mod.rs

```diff
@@ -101,9 +109,46 @@ impl<S: BlockSigner + Send + Sync + 'static> api_server::Api for ValidatorServer
     /// Receives a proven transaction, then validates and stores it.
     async fn submit_proven_transaction(
         &self,
-        _request: tonic::Request<proto::transaction::ProvenTransaction>,
+        request: tonic::Request<proto::transaction::ProvenTransaction>,
     ) -> Result<tonic::Response<()>, tonic::Status> {
-        // TODO(sergerad): Implement transaction validation logic.
+        let request = request.into_inner();
+        // Deserialize the transaction.
+        let proven_tx =
+            ProvenTransaction::read_from_bytes(&request.transaction).map_err(|err| {
+                Status::invalid_argument(err.as_report_context("Invalid proven transaction"))
+            })?;
+
+        // Deserialize the transaction inputs.
+        let Some(tx_inputs) = request.transaction_inputs else {
+            return Err(Status::invalid_argument("Missing transaction inputs"));
+        };
+        let tx_inputs = TransactionInputs::read_from_bytes(&tx_inputs).map_err(|err| {
+            Status::invalid_argument(err.as_report_context("Invalid transaction inputs"))
+        })?;
+
+        // Create a DataStore from the transaction inputs.
+        let data_store = TransactionInputsDataStore::new(tx_inputs.clone());
+
+        // Execute the transaction.
+        let (account, block_header, _, input_notes, tx_args) = tx_inputs.into_parts();
+        let executor: TransactionExecutor<'_, '_, _, UnreachableAuth> =
+            TransactionExecutor::new(&data_store);
+        let executed_tx = executor
+            .execute_transaction(account.id(), block_header.block_num(), input_notes, tx_args)
+            .await
+            .map_err(|err| {
+                Status::internal(err.as_report_context("Failed to execute transaction"))
```

Comment:
I'm not sure if the error status here should be "internal". This could be the result of wrong transaction inputs? So, I'd treat it as input error.